artifactory/commands/utils/interactive_questionnaire.go

```diff
@@ -0,0 +1,247 @@
+type questionCallback func(*InteractiveQuestionnaire, string) (string, error)
+
+type QuestionInfo struct {
+	Msg          string
+	PromptPrefix string
+	Options      []prompt.Suggest
+	AllowVars    bool
+	Writer       AnswerWriter
+	MapKey       string
+	Callback     questionCallback
+}
+
+const (
+	PressTabMsg      = " (press Tab for options):"
+	InvalidAnswerMsg = "Invalid answer. Please select value from the suggestions list."
+	VariableUseMsg   = " You may use dynamic variable in the form of ${key}."
+	EmptyValueMsg    = "The value can not be empty. Please enter a valid value:"
+	OptionalKey      = "OptionalKey"
+	WriteAndExist    = ":x"
+
+	// Boolean answers
+	True  = "true"
+	False = "false"
+
+	CommaSeparatedListMsg = "The value should be a comma separated list"
+)
+
+// Var can be inserted in the form of ${key}
+var VarPattern = regexp.MustCompile(`^\$\{\w+\}+$`)
+
+func prefixCompleter(options []prompt.Suggest) prompt.Completer {
+	return func(document prompt.Document) []prompt.Suggest {
+		return prompt.FilterHasPrefix(options, document.GetWordBeforeCursor(), true)
+	}
+}
+
+// Ask question with free string answer, answer cannot be empty.
+// Variable aren't check and can be part of the answer
+func AskString(msg, promptPrefix string) string {
+	if msg != "" {
+		fmt.Println(msg + ":")
+	}
+	for {
+		answer := prompt.Input(promptPrefix+" ", prefixCompleter(nil))
+		if answer != "" {
+			return answer
+		}
+		fmt.Println(EmptyValueMsg)
+	}
+}
+
+// Ask question with list of possible answers.
+// The answer must be chosen from the list, but can be a variable if allowVars set to true.
+func AskFromList(msg, promptPrefix string, allowVars bool, options []prompt.Suggest) string {
+	if msg != "" {
+		fmt.Println(msg + PressTabMsg)
+	}
+	errMsg := InvalidAnswerMsg
+	if allowVars {
+		errMsg += VariableUseMsg
+	}
+	for {
+		answer := prompt.Input(promptPrefix+" ", prefixCompleter(options))
+		if validateAnswer(answer, options, allowVars) {
+			return answer
+		}
+		fmt.Println(errMsg)
+	}
+}
+
+func validateAnswer(answer string, options []prompt.Suggest, allowVars bool) bool {
+	if allowVars {
+		if regexMatch := VarPattern.FindStringSubmatch(answer); regexMatch != nil {
+			return true
+		}
+	}
+	for _, option := range options {
+		if answer == option.Text {
+			return true
+		}
+	}
+	return false
+}
+
+// Ask question steps:
+// 		1. Ask for string/from list
+//		2. Write the answer to answersMap (if writer provided)
+// 		3. Run callback (if provided)q
+func (iq *InteractiveQuestionnaire) AskQuestion(question QuestionInfo) (value string, err error) {
+
+	var answer string
+	if question.Options != nil {
+		answer = AskFromList(question.Msg, question.PromptPrefix, question.AllowVars, question.Options)
+	} else {
+		answer = AskString(question.Msg, question.PromptPrefix)
+	}
+	if question.Writer != nil {
+		//err = question.Writer(&iq.AnswersMap, question.MapKey, answer)
```

Comment:
This comment should be probably removed.

Reply:
Actually this is the right line should stay -
We want to keep this questionnaire generic as possible, so we use the writer and not write every answer as string.
I added new writers at template.go that writes every question as a string.